lib/compiler/src/artifact_builders/artifact_builder.rs

```diff
@@ -232,13 +244,234 @@ impl ArtifactCreate for ArtifactBuild {
     }
 
     fn serialize(&self) -> Result<Vec<u8>, SerializeError> {
-        let serialized_data = self.serializable.serialize()?;
-        assert!(std::mem::align_of::<SerializableModule>() <= MetadataHeader::ALIGN);
+        serialize_module(&self.serializable)
+    }
+}
+
+/// Module loaded from an archive. Since `CompileModuleInfo` is part of the public
+/// interface of this crate and has to be mutable, it has to be deserialized completely.
+pub struct ModuleFromArchive<'a> {
+    /// The main serializable compilation object
+    pub compilation: &'a ArchivedSerializableCompilation,
+    /// Datas initializers
+    pub data_initializers: &'a rkyv::Archived<Box<[OwnedDataInitializer]>>,
+    /// CPU Feature flags for this compilation
+    pub cpu_features: u64,
+
+    // Keep the original module around for re-serialization
+    original_module: &'a ArchivedSerializableModule,
+}
+
+impl<'a> ModuleFromArchive<'a> {
+    /// Create a new `ModuleFromArchive` from the archived version of a `SerializableModule`
+    pub fn from_serializable_module(
+        module: &'a ArchivedSerializableModule,
+    ) -> Result<Self, DeserializeError> {
+        Ok(Self {
+            compilation: &module.compilation,
+            data_initializers: &module.data_initializers,
+            cpu_features: module.cpu_features,
+            original_module: module,
+        })
+    }
+}
+
+self_cell!(
+    struct ArtifactBuildFromArchiveCell {
+        owner: OwnedBuffer,
+
+        #[covariant]
+        dependent: ModuleFromArchive,
+    }
+);
+
+/// A compiled wasm module that was loaded from a serialized archive.
+pub struct ArtifactBuildFromArchive {
+    cell: ArtifactBuildFromArchiveCell,
+
+    /// Compilation informations
+    compile_info: CompileModuleInfo,
+}
+
+impl ArtifactBuildFromArchive {
+    pub(crate) fn try_new(
+        buffer: OwnedBuffer,
+        module_builder: impl FnOnce(
+            &OwnedBuffer,
+        ) -> Result<&ArchivedSerializableModule, DeserializeError>,
+    ) -> Result<Self, DeserializeError> {
+        let mut compile_info = MaybeUninit::uninit();
+
+        let cell = ArtifactBuildFromArchiveCell::try_new(buffer, |buffer| {
+            let module = module_builder(buffer)?;
+            let mut deserializer = SharedDeserializeMap::new();
+            compile_info = MaybeUninit::new(
+                rkyv::Deserialize::deserialize(&module.compile_info, &mut deserializer)
+                    .map_err(|e| DeserializeError::CorruptedBinary(format!("{:?}", e)))?,
+            );
+            ModuleFromArchive::from_serializable_module(module)
+        })?;
+
+        // Safety: we know the lambda will execute before getting here and assign both values
+        Ok(Self {
+            cell,
+            compile_info: unsafe { compile_info.assume_init() },
+        })
```

Comment:
Nit: for better readability, can we do:

```suggestion
        // Safety: we know the lambda will execute before getting here and assign both values
        let compile_info = unsafe { compile_info.assume_init() };
        Ok(Self {
            cell,
            compile_info,
        })
```